service/src/main/java/org/mskcc/smile/service/impl/DbGapMessageHandlingServiceImpl.java

```diff
@@ -0,0 +1,170 @@
+import io.nats.client.Message;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.mskcc.cmo.messaging.Gateway;
+import org.mskcc.cmo.messaging.MessageConsumer;
+import org.mskcc.smile.model.SmileSample;
+import org.mskcc.smile.model.json.DbGapJson;
+import org.mskcc.smile.service.DbGapMessageHandlingService;
+import org.mskcc.smile.service.DbGapService;
+import org.mskcc.smile.service.SmileSampleService;
+import org.mskcc.smile.service.util.NatsMsgUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DbGapMessageHandlingServiceImpl implements DbGapMessageHandlingService {
+    @Value("${dbgap.sample_update_topic}")
+    private String DBGAP_SAMPLE_UPDATE_TOPIC;
+
+    @Value("${num.dbgap_msg_handler_threads}")
+    private int NUM_DBGAP_MSG_HANDLERS;
+
+    @Autowired
+    private DbGapService dbGapService;
+
+    @Autowired
+    private SmileSampleService sampleService;
+
+    private static Gateway messagingGateway;
+    private static final Log LOG = LogFactory.getLog(DbGapMessageHandlingServiceImpl.class);
+
+    private static boolean initialized = false;
+    private static volatile boolean shutdownInitiated;
+    private static final ExecutorService exec = Executors.newCachedThreadPool();
+
+    private static final BlockingQueue<DbGapJson> dbGapUpdateQueue =
+            new LinkedBlockingQueue<DbGapJson>();
+
+    private static CountDownLatch dbgapUpdateHandlerShutdownLatch;
+
+    private class DbGapUpdateHandler implements Runnable {
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        DbGapUpdateHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    DbGapJson dbGap = dbGapUpdateQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (dbGap != null) {
+                        // this message is coming straight from the dashboard and should therefore always
+                        // have data for a valid sample that exists in the database
+                        // however this check will make extra sure that the primary id received
+                        // in the nats message actually exists in the database before conducting
+                        // further operations in the db
+                        SmileSample sample = sampleService.getDetailedSampleByInputId(dbGap.getPrimaryId());
+                        if (sample != null) {
```

Comment:
Copied this safety check from the sample billing update handler [here](https://github.com/mskcc/smile-server/blob/master/service/src/main/java/org/mskcc/smile/service/impl/TempoMessageHandlingServiceImpl.java#L345). I think this makes sense also for cases where we make manual updates to DbGaps via a Python script or publisher tool.